model/flow/identity.go

```diff
@@ -476,20 +487,63 @@ func (il IdentityList) SamplePct(pct float64) IdentityList {
 // Union returns a new identity list containing every identity that occurs in
 // either `il`, or `other`, or both. There are no duplicates in the output,
 // where duplicates are identities with the same node ID.
+// The returned IdentityList is sorted
 func (il IdentityList) Union(other IdentityList) IdentityList {
-
 	// stores the output, the union of the two lists
-	union := make(IdentityList, 0, len(il)+len(other))
-	// efficient lookup to avoid duplicates
-	lookup := make(map[Identifier]struct{})
-
-	// add all identities, omitted duplicates
-	for _, identity := range append(il.Copy(), other...) {
-		if _, exists := lookup[identity.NodeID]; exists {
-			continue
+	if (len(il) + len(other)) == 0 {
+		return IdentityList{}
+	}
+
+	// add all identities together
+	union := append(other, il...)
```

Comment:
That would kind of nullify the reason for this change...  We're seeing excessive garbage collection activity, and this is one of the top contributors.

As an alternative, how about we require the 'other IdentityList' to be 'other *IdentityList' so that the developers see that there is something different, and that the value is mutable, plus add extensive comments ?  I checked the code that calls Union, and the calling spots are discarding 'other' after it's used.

Reply:
Overall, we prioritize safety, for the protocol as well as the implementation. Making some performance compromises to reduce the possibility to miss-use a function is the preferred choice. 

Instead of modifying the default implementation to be unsafe, and provide a safer alternative, you could do it the other way around:
* `IdentityList.Union` is safe by not modifying the inputs
* For the few performance-critical places in the code, you could provide an unsafe alternative which works in-place and clearly distinguish it by name, i.e. `IdentityList.UnsafeUnionInPlace`. Thereby developers will chose the by default the safe option. 

Regarding your argument that copying would defeat the purpose of the optimization, I am not really convinced: 
* Lets look at your second example https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L81 specifically the place in the code where `topicFanout` is created: https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L77
*  going into the function `subsetChannel`, you'll find that there are a lot of `Filter` function applied ([here](https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L98, [here](https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L137-L138), ...) and finally we apply the [`sampleFanout`](https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L112) function. What both `sampleFanout` and  `Filter` have in common is that (for safety reasons), they return a new slice: https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/randomizedTopology.go#L117 https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/model/flow/identity.go#L290
* After closer inspection of the mechanics of these function, we learn that these function will already copy the content multiple times (internally via the append function). 

So in summary, there are many re-allocations through the `sampleFanout` and  `Filter` functions anyway. For the sake of argument, lets just assume that these function re-allocate in total 10 new arrays. The safe implementation of `Union` does a _single_ new allocation (we specifically avoid append). I really don't see why removing this last allocation in `Union` would make a big difference.

Reply:
BTW, here is the reason why I am not ok with in-place modifications for the `IdentityList` methods (except for explicitly distinguished unsafe alternatives): 
* Here we have the identity lists for the current Epoch: https://github.com/onflow/flow-go/blob/03cba83741a4e43e34f64761147f98728480b1ff/network/topology/topology_utils.go#L55 All Identity filters ultimately use this list as their basis. If you happen to write through to this array (or modify the code such that some other place in the code will modify this list), it is game over. If that happens only once, you will have completely broken Flow:
  * then, the nodes might reject legitimate messages from other nodes  
  * or (oven worse) they might accept messages from malicious nodes that are not authorized to send such messages 

This is not a risk I am willing to take. Flow is already complex enough. We can't have engineers carefully work around subtle unsafe implementations in addition. APIs need to be as safe as possible. Making broadly-used components unsafe, because it benefits a few specific spots in the code is not an option. If you need highly performant implementations at the cost of API safety, please provide a special-purpose implementation that is clearly distinguished as unsafe.

Reply:
I agree that it's ugly.  Most performance optimizations are.

Seniority wins.  I'll shelve the Union performance fix.